File: cpp/Platform.Collections.Methods.Tests/TestExtensions.cpp
﻿namespace Platform::Collections::Methods::Tests
{
    class Assert
    {
    public:
        template<typename T1, typename T2>
        static void AreEqual(T1 a, T2 b)
        {
            if (a != b)
            {
                throw std::runtime_error("Values are not equal");
            }
        }
    };
    
    class TestExtensions
    {
        public: template <typename impl_t, typename TElement> static void TestMultipleCreationsAndDeletions(impl_t& tree, auto allocate, auto free, TElement* root, auto treeCount, std::int32_t maximumOperationsPerCycle)
        {
            for (auto N = 1; N < maximumOperationsPerCycle; N++)
            {
                auto currentCount = 0;
                for (auto i = 0; i < N; i++)
                {
                    auto node = allocate();
                    tree.Attach(root, node);
                    currentCount++;
                    Assert::AreEqual(currentCount, (std::int32_t)treeCount());
                }
                for (auto i = 1; i <= N; i++)
                {
                    TElement node = i;
                    if (tree.Contains(node, *root))
                    {
                        tree.Detach(root, node);
                        free(node);
                        currentCount--;
                        Assert::AreEqual(currentCount, (std::int32_t)treeCount());
                    }
                }
            }
        }

        public: template <typename impl_t, typename TElement> static void TestMultipleRandomCreationsAndDeletions(impl_t& tree, TElement* root, auto treeCount, std::int32_t maximumOperationsPerCycle)
        {
            std::srand(0);
            std::unordered_set<TElement> added;
            auto currentCount = 0;
            for (auto N = 1; N < maximumOperationsPerCycle; N++)
            {
                for (auto i = 0; i < N; i++)
                {
                    auto node = (std::rand() % N) + 1;
                    if (!added.contains(node))
                    {
                        added.insert(node);
                        tree.Attach(root, node);
                        currentCount++;
                        Assert::AreEqual(currentCount, (std::int32_t)treeCount());
                    }
                }
                for (auto i = 1; i <= N; i++)
                {
                    TElement node = (std::rand() % N) + 1;
                    if (tree.Contains(node, *root))
                    {
                        tree.Detach(root, node);
                        currentCount--;
                        Assert::AreEqual(currentCount, (std::int32_t)treeCount());
                        added.erase(node);
                    }
                }
            }
        }
    };
}

File: csharp/Platform.Collections.Methods.Tests/TestExtensions.cs
using System;
using System.Collections.Generic;
using Xunit;
using Platform.Collections.Methods.Trees;
using Platform.Converters;

namespace Platform.Collections.Methods.Tests
{
    /// <summary>
    /// <para>
    /// Represents the test extensions.
    /// </para>
    /// <para></para>
    /// </summary>
    public static class TestExtensions
    {
        /// <summary>
        /// <para>
        /// Tests the multiple creations and deletions using the specified tree.
        /// </para>
        /// <para></para>
        /// </summary>
        /// <typeparam name="TElement">
        /// <para>The element.</para>
        /// <para></para>
        /// </typeparam>
        /// <param name="tree">
        /// <para>The tree.</para>
        /// <para></para>
        /// </param>
        /// <param name="allocate">
        /// <para>The allocate.</para>
        /// <para></para>
        /// </param>
        /// <param name="free">
        /// <para>The free.</para>
        /// <para></para>
        /// </param>
        /// <param name="root">
        /// <para>The root.</para>
        /// <para></para>
        /// </param>
        /// <param name="treeCount">
        /// <para>The tree count.</para>
        /// <para></para>
        /// </param>
        /// <param name="maximumOperationsPerCycle">
        /// <para>The maximum operations per cycle.</para>
        /// <para></para>
        /// </param>
        public static void TestMultipleCreationsAndDeletions<TElement>(this SizedBinaryTreeMethodsBase<TElement> tree, Func<TElement> allocate, Action<TElement> free, ref TElement root, Func<TElement> treeCount, int maximumOperationsPerCycle)
        {
            for (var N = 1; N < maximumOperationsPerCycle; N++)
            {
                var currentCount = 0;
                for (var i = 0; i < N; i++)
                {
                    var node = allocate();
                    tree.Attach(ref root, node);
                    currentCount++;
                    Assert.Equal(currentCount, (int)UncheckedConverter<TElement, int>.Default.Convert(treeCount()));
                }
                for (var i = 1; i <= N; i++)
                {
                    TElement node = UncheckedConverter<int, TElement>.Default.Convert(i);
                    if (tree.Contains(node, root))
                    {
                        tree.Detach(ref root, node);
                        free(node);
                        currentCount--;
                        Assert.Equal(currentCount, (int)UncheckedConverter<TElement, int>.Default.Convert(treeCount()));
                    }
                }
            }
        }

        /// <summary>
        /// <para>
        /// Tests the multiple random creations and deletions using the specified tree.
        /// </para>
        /// <para></para>
        /// </summary>
        /// <typeparam name="TElement">
        /// <para>The element.</para>
        /// <para></para>
        /// </typeparam>
        /// <param name="tree">
        /// <para>The tree.</para>
        /// <para></para>
        /// </param>
        /// <param name="root">
        /// <para>The root.</para>
        /// <para></para>
        /// </param>
        /// <param name="treeCount">
        /// <para>The tree count.</para>
        /// <para></para>
        /// </param>
        /// <param name="maximumOperationsPerCycle">
        /// <para>The maximum operations per cycle.</para>
        /// <para></para>
        /// </param>
        public static void TestMultipleRandomCreationsAndDeletions<TElement>(this SizedBinaryTreeMethodsBase<TElement> tree, ref TElement root, Func<TElement> treeCount, int maximumOperationsPerCycle)
        {
            var random = new System.Random(0);
            var added = new HashSet<TElement>();
            var currentCount = 0;
            for (var N = 1; N < maximumOperationsPerCycle; N++)
            {
                for (var i = 0; i < N; i++)
                {
                    var node = UncheckedConverter<int, TElement>.Default.Convert(random.Next(1, N));
                    if (added.Add(node))
                    {
                        tree.Attach(ref root, node);
                        currentCount++;
                        Assert.Equal(currentCount, (int)UncheckedConverter<TElement, int>.Default.Convert(treeCount()));
                    }
                }
                for (var i = 1; i <= N; i++)
                {
                    TElement node = UncheckedConverter<int, TElement>.Default.Convert(random.Next(1, N));
                    if (tree.Contains(node, root))
                    {
                        tree.Detach(ref root, node);
                        currentCount--;
                        Assert.Equal(currentCount, (int)UncheckedConverter<TElement, int>.Default.Convert(treeCount()));
                        added.Remove(node);
                    }
                }
            }
        }
    }
}
